// drizzle-orm/bun-sql/mysql/session.js
import { entityKind, is } from "../../entity.js";
import { Column } from "../../column.js";
import { mapResultRow } from "../../utils.js";
import { fillPlaceholders } from "../../sql/sql.js";
import { NoopLogger } from "../../logger.js";
import { NoopCache } from "../../cache/core/index.js";
import { MySqlPreparedQuery, MySqlSession, MySqlTransaction } from "../../mysql-core/session.js";

//#region src/bun-sql/mysql/session.ts
var BunMySqlPreparedQuery = class extends MySqlPreparedQuery {
	static [entityKind] = "BunMySqlPreparedQuery";
	constructor(client, query, logger, cache, queryMetadata, cacheConfig, fields, customResultMapper, generatedIds, returningIds, isRqbV2Query) {
		super(query, cache, queryMetadata, cacheConfig);
		this.client = client;
		this.logger = logger;
		this.fields = fields;
		this.customResultMapper = customResultMapper;
		this.generatedIds = generatedIds;
		this.returningIds = returningIds;
		this.isRqbV2Query = isRqbV2Query;
	}
	async execute(placeholderValues = {}) {
		if (this.isRqbV2Query) return this.executeRqbV2(placeholderValues);
		const { fields, client, logger, query, joinsNotNullableMap, customResultMapper, returningIds, generatedIds } = this;
		const params = fillPlaceholders(query.params, placeholderValues);
		logger.logQuery(query.sql, params);
		if (!fields && !customResultMapper) {
			const res = await this.queryWithCache(query.sql, params, async () => {
				return await client.unsafe(query.sql, params);
			});
			const insertId = res.lastInsertRowid;
			const affectedRows = res.affectedRows;
			if (returningIds) {
				const returningResponse = [];
				let j = 0;
				for (let i = insertId; i < insertId + affectedRows; i++) {
					for (const column of returningIds) {
						const key = returningIds[0].path[0];
						if (is(column.field, Column)) {
							if (column.field.primary && column.field.autoIncrement) returningResponse.push({ [key]: i });
							if (column.field.defaultFn && generatedIds) returningResponse.push({ [key]: generatedIds[j][key] });
						}
					}
					j++;
				}
				return returningResponse;
			}
			return res;
		}
		const rows = await this.queryWithCache(query.sql, params, async () => {
			return await client.unsafe(query.sql, params).values();
		});
		if (customResultMapper) return customResultMapper(rows);
		return rows.map((row) => mapResultRow(fields, row, joinsNotNullableMap));
	}
	async executeRqbV2(placeholderValues = {}) {
		const params = fillPlaceholders(this.query.params, placeholderValues);
		this.logger.logQuery(this.query.sql, params);
		const { client, query, customResultMapper } = this;
		return customResultMapper(await client.unsafe(query.sql, params));
	}
	async *iterator(placeholderValues = {}) {
		const { fields, query, joinsNotNullableMap, client, customResultMapper } = this;
		const params = fillPlaceholders(query.params, placeholderValues);
		const hasRowsMapper = Boolean(fields || customResultMapper);
		const rows = await this.queryWithCache(query.sql, params, async () => {
			return await (hasRowsMapper && !this.isRqbV2Query ? client.unsafe(query.sql, params).values() : client.unsafe(query.sql, params));
		});
		for (const row of rows) {
			if (row === void 0 || Array.isArray(row) && row.length === 0) break;
			if (hasRowsMapper) if (customResultMapper) {
				const mappedRow = customResultMapper([row]);
				yield Array.isArray(mappedRow) ? mappedRow[0] : mappedRow;
			} else yield mapResultRow(fields, row, joinsNotNullableMap);
			else yield row;
		}
	}
};
var BunMySqlSession = class BunMySqlSession extends MySqlSession {
	static [entityKind] = "BunMySqlSession";
	logger;
	mode;
	cache;
	constructor(client, dialect, relations, schema, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.schema = schema;
		this.options = options;
		this.logger = options.logger ?? new NoopLogger();
		this.cache = options.cache ?? new NoopCache();
		this.mode = options.mode;
	}
	prepareQuery(query, fields, customResultMapper, generatedIds, returningIds, queryMetadata, cacheConfig) {
		return new BunMySqlPreparedQuery(this.client, query, this.logger, this.cache, queryMetadata, cacheConfig, fields, customResultMapper, generatedIds, returningIds);
	}
	prepareRelationalQuery(query, fields, customResultMapper, generatedIds, returningIds) {
		return new BunMySqlPreparedQuery(this.client, query, this.logger, this.cache, void 0, void 0, fields, customResultMapper, generatedIds, returningIds, true);
	}
	/** @internal */
	async query(query, params) {
		this.logger.logQuery(query, params);
		return await this.client.unsafe(query, params);
	}
	all(query) {
		const querySql = this.dialect.sqlToQuery(query);
		this.logger.logQuery(querySql.sql, querySql.params);
		return this.client.unsafe(querySql.sql, querySql.params);
	}
	async count(sql) {
		const query = this.dialect.sqlToQuery(sql);
		const count = (await this.client.unsafe(query.sql, query.params).values())[0][0];
		if (typeof count === "number") return count;
		return Number(count);
	}
	async transaction(transaction, config) {
		const startTransactionSql = config ? this.getStartTransactionSQL(config)?.inlineParams().toQuery(this.dialect).sql.slice(18) ?? "" : "";
		if (config?.isolationLevel) throw new Error("Driver doesn't support setting isolation level on transaction");
		return this.client.begin(startTransactionSql, async (client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.schema, this.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, this.schema, 0, this.mode));
		});
	}
};
var BunMySqlTransaction = class BunMySqlTransaction extends MySqlTransaction {
	static [entityKind] = "BunMySqlTransaction";
	async transaction(transaction) {
		return this.session.client.savepoint((client) => {
			const session = new BunMySqlSession(client, this.dialect, this.relations, this.schema, this.session.options);
			return transaction(new BunMySqlTransaction(this.dialect, session, this.relations, this.schema, this.nestedIndex + 1, this.mode));
		});
	}
};

//#endregion
export { BunMySqlPreparedQuery, BunMySqlSession, BunMySqlTransaction };
