// drizzle-orm/pg-core/async/update.js
import { extractUsedTable } from "../utils.js";
import { PgUpdateBase } from "../query-builders/update.js";
import { entityKind } from "../../entity.js";
import { applyMixins } from "../../utils.js";
import { QueryPromise } from "../../query-promise.js";
import { preparedStatementName } from "../../query-name-generator.js";

//#region src/pg-core/async/update.ts
var PgAsyncUpdateBase = class extends PgUpdateBase {
	static [entityKind] = "PgAsyncUpdate";
	/** @internal */
	_prepare(name, generateName = false) {
		const query = this.dialect.sqlToQuery(this.getSQL());
		const preparedQuery = this.session.prepareQuery(query, this.config.returning, name ?? (generateName ? preparedStatementName(query.sql, query.params) : name), true, void 0, {
			type: "insert",
			tables: extractUsedTable(this.config.table)
		}, this.cacheConfig);
		preparedQuery.joinsNotNullableMap = this.joinsNotNullableMap;
		return preparedQuery.setToken(this.authToken);
	}
	prepare(name) {
		return this._prepare(name, true);
	}
	/** @internal */
	authToken;
	/** @internal */
	setToken(token) {
		this.authToken = token;
		return this;
	}
	execute = (placeholderValues = {}) => {
		return this._prepare().execute(placeholderValues);
	};
};
applyMixins(PgAsyncUpdateBase, [QueryPromise]);

//#endregion
export { PgAsyncUpdateBase };
