// drizzle-orm/pg-core/async/update.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_pg_core_utils = require('../utils.cjs');
const require_pg_core_query_builders_update = require('../query-builders/update.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __query_promise_ts = require("../../query-promise.cjs");
let __query_name_generator_ts = require("../../query-name-generator.cjs");

//#region src/pg-core/async/update.ts
var PgAsyncUpdateBase = class extends require_pg_core_query_builders_update.PgUpdateBase {
	static [__entity_ts.entityKind] = "PgAsyncUpdate";
	/** @internal */
	_prepare(name, generateName = false) {
		const query = this.dialect.sqlToQuery(this.getSQL());
		const preparedQuery = this.session.prepareQuery(query, this.config.returning, name ?? (generateName ? (0, __query_name_generator_ts.preparedStatementName)(query.sql, query.params) : name), true, void 0, {
			type: "insert",
			tables: require_pg_core_utils.extractUsedTable(this.config.table)
		}, this.cacheConfig);
		preparedQuery.joinsNotNullableMap = this.joinsNotNullableMap;
		return preparedQuery.setToken(this.authToken);
	}
	prepare(name) {
		return this._prepare(name, true);
	}
	/** @internal */
	authToken;
	/** @internal */
	setToken(token) {
		this.authToken = token;
		return this;
	}
	execute = (placeholderValues = {}) => {
		return this._prepare().execute(placeholderValues);
	};
};
(0, __utils_ts.applyMixins)(PgAsyncUpdateBase, [__query_promise_ts.QueryPromise]);

//#endregion
exports.PgAsyncUpdateBase = PgAsyncUpdateBase;
