// drizzle-orm/netlify-db/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __logger_ts = require("../logger.cjs");
let __pg_core_async_session_ts = require("../pg-core/async/session.cjs");
let __cache_core_cache_ts = require("../cache/core/cache.cjs");
let _neondatabase_serverless = require("@neondatabase/serverless");
let __neon_http_session_ts = require("../neon-http/session.cjs");
let __neon_serverless_session_ts = require("../neon-serverless/session.cjs");

//#region src/netlify-db/session.ts
/**
* Ensures a WebSocket implementation is available for Neon's
* serverless driver. The Neon driver checks for a global WebSocket
* automatically, but this sets it on neonConfig explicitly so the
* check only happens once.
*/
function ensureWebSocket() {
	if (_neondatabase_serverless.neonConfig.webSocketConstructor) return;
	if (typeof WebSocket !== "undefined") _neondatabase_serverless.neonConfig.webSocketConstructor = WebSocket;
}
const queryConfig = {
	arrayMode: true,
	fullResults: true
};
var NetlifyDbSession = class extends __pg_core_async_session_ts.PgAsyncSession {
	static [__entity_ts.entityKind] = "NetlifyDbSession";
	clientQuery;
	logger;
	cache;
	constructor(httpClient, pool, dialect, relations, schema, options = {}) {
		super(dialect);
		this.httpClient = httpClient;
		this.pool = pool;
		this.relations = relations;
		this.schema = schema;
		this.options = options;
		this.clientQuery = httpClient.query ?? httpClient;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_cache_ts.NoopCache();
	}
	prepareQuery(query, fields, name, isResponseInArrayMode, customResultMapper, queryMetadata, cacheConfig) {
		return new __neon_http_session_ts.NeonHttpPreparedQuery(this.httpClient, query, this.logger, this.cache, queryMetadata, cacheConfig, fields, isResponseInArrayMode, customResultMapper);
	}
	prepareRelationalQuery(query, fields, name, customResultMapper) {
		return new __neon_http_session_ts.NeonHttpPreparedQuery(this.httpClient, query, this.logger, this.cache, void 0, void 0, fields, false, customResultMapper, true);
	}
	async batch(queries) {
		const preparedQueries = [];
		const builtQueries = [];
		for (const query of queries) {
			const preparedQuery = query._prepare();
			const builtQuery = preparedQuery.getQuery();
			preparedQueries.push(preparedQuery);
			builtQueries.push(this.clientQuery(builtQuery.sql, builtQuery.params, {
				fullResults: true,
				arrayMode: preparedQuery.isResponseInArrayMode()
			}));
		}
		return (await this.httpClient.transaction(builtQueries, queryConfig)).map((result, i) => preparedQueries[i].mapResult(result, true));
	}
	async query(query, params) {
		this.logger.logQuery(query, params);
		return await this.clientQuery(query, params, {
			arrayMode: true,
			fullResults: true
		});
	}
	async queryObjects(query, params) {
		return this.clientQuery(query, params, {
			arrayMode: false,
			fullResults: true
		});
	}
	async transaction(transaction, config = {}) {
		ensureWebSocket();
		const poolClient = await this.pool.connect();
		const session = new NetlifyDbWsSession(poolClient, this.dialect, this.relations, this.schema, this.options);
		const tx = new NetlifyDbTransaction(this.dialect, session, this.relations, this.schema);
		await tx.execute(__sql_sql_ts.sql`begin ${tx.getTransactionConfigSQL(config)}`);
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql`commit`);
			return result;
		} catch (error) {
			await tx.execute(__sql_sql_ts.sql`rollback`);
			throw error;
		} finally {
			poolClient.release();
		}
	}
};
/**
* Internal WebSocket-based session used only within transactions.
* Delegates all queries to a PoolClient over WebSocket, using
* NeonPreparedQuery from the neon-serverless adapter.
*/
var NetlifyDbWsSession = class extends __pg_core_async_session_ts.PgAsyncSession {
	static [__entity_ts.entityKind] = "NetlifyDbWsSession";
	logger;
	cache;
	constructor(client, dialect, relations, schema, options = {}) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.schema = schema;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_cache_ts.NoopCache();
	}
	prepareQuery(query, fields, name, isResponseInArrayMode, customResultMapper, queryMetadata, cacheConfig) {
		return new __neon_serverless_session_ts.NeonPreparedQuery(this.client, query, this.logger, this.cache, queryMetadata, cacheConfig, fields, name, isResponseInArrayMode, customResultMapper);
	}
	prepareRelationalQuery(query, fields, name, customResultMapper) {
		return new __neon_serverless_session_ts.NeonPreparedQuery(this.client, query, this.logger, this.cache, void 0, void 0, fields, name, false, customResultMapper, true);
	}
	async query(query, params) {
		this.logger.logQuery(query, params);
		return await this.client.query({
			rowMode: "array",
			text: query,
			values: params
		});
	}
	async queryObjects(query, params) {
		return this.client.query(query, params);
	}
	async transaction(_transaction, _config) {
		throw new Error("Nested transactions are handled by NetlifyDbTransaction via savepoints");
	}
};
var NetlifyDbTransaction = class NetlifyDbTransaction extends __pg_core_async_session_ts.PgAsyncTransaction {
	static [__entity_ts.entityKind] = "NetlifyDbTransaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex + 1}`;
		const tx = new NetlifyDbTransaction(this.dialect, this.session, this.relations, this.schema, this.nestedIndex + 1);
		await tx.execute(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await tx.execute(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (e) {
			await tx.execute(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw e;
		}
	}
};

//#endregion
exports.NetlifyDbSession = NetlifyDbSession;
exports.NetlifyDbTransaction = NetlifyDbTransaction;
exports.NetlifyDbWsSession = NetlifyDbWsSession;
