// drizzle-orm/effect-postgres/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __utils_ts = require("../utils.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __effect_core_errors_ts = require("../effect-core/errors.cjs");
let __effect_core_logger_ts = require("../effect-core/logger.cjs");
let __pg_core_effect_session_ts = require("../pg-core/effect/session.cjs");

//#region src/effect-postgres/session.ts
var EffectPgPreparedQuery = class extends __pg_core_effect_session_ts.PgEffectPreparedQuery {
	static [__entity_ts.entityKind] = "EffectPgPreparedQuery";
	constructor(client, query, logger, cache, queryMetadata, cacheConfig, fields, name, _isResponseInArrayMode, customResultMapper, isRqbV2Query) {
		super(query, cache, queryMetadata, cacheConfig);
		this.client = client;
		this.logger = logger;
		this.fields = fields;
		this._isResponseInArrayMode = _isResponseInArrayMode;
		this.customResultMapper = customResultMapper;
		this.isRqbV2Query = isRqbV2Query;
	}
	execute(placeholderValues) {
		return effect_Effect.gen(this, function* () {
			if (this.isRqbV2Query) return yield* this.executeRqbV2(placeholderValues);
			const { query, customResultMapper, fields, joinsNotNullableMap, client } = this;
			const params = (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues ?? {});
			yield* __effect_core_logger_ts.EffectLogger.logQuery(query.sql, params);
			if (!fields && !customResultMapper) return yield* this.queryWithCache(query.sql, params, this.client.unsafe(query.sql, params).withoutTransform);
			return yield* this.queryWithCache(query.sql, params, client.unsafe(query.sql, params).values).pipe(effect_Effect.map((rows) => {
				if (customResultMapper) return customResultMapper(rows);
				return rows.map((row) => (0, __utils_ts.mapResultRow)(fields, row, joinsNotNullableMap));
			}));
		}).pipe(effect_Effect.provideService(__effect_core_logger_ts.EffectLogger, this.logger));
	}
	executeRqbV2(placeholderValues) {
		return effect_Effect.gen(this, function* () {
			const { query, customResultMapper, client } = this;
			const params = (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues ?? {});
			yield* __effect_core_logger_ts.EffectLogger.logQuery(query.sql, params);
			return yield* client.unsafe(query.sql, params).withoutTransform.pipe(effect_Effect.flatMap((v) => effect_Effect.try(() => customResultMapper(v))), effect_Effect.catchAll((e) => new __effect_core_errors_ts.EffectDrizzleQueryError({
				query: query.sql,
				params,
				cause: e
			})));
		}).pipe(effect_Effect.provideService(__effect_core_logger_ts.EffectLogger, this.logger));
	}
	all(placeholderValues) {
		return effect_Effect.gen(this, function* () {
			const { query, client } = this;
			const params = (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues ?? {});
			yield* __effect_core_logger_ts.EffectLogger.logQuery(query.sql, params);
			return yield* this.queryWithCache(query.sql, params, client.unsafe(query.sql, params).withoutTransform);
		}).pipe(effect_Effect.provideService(__effect_core_logger_ts.EffectLogger, this.logger));
	}
	/** @internal */
	isResponseInArrayMode() {
		return this._isResponseInArrayMode;
	}
};
var EffectPgSession = class extends __pg_core_effect_session_ts.PgEffectSession {
	static [__entity_ts.entityKind] = "EffectPgSession";
	constructor(client, dialect, relations, schema, logger, cache) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.schema = schema;
		this.logger = logger;
		this.cache = cache;
	}
	prepareQuery(query, fields, name, isResponseInArrayMode, customResultMapper, queryMetadata, cacheConfig) {
		return new EffectPgPreparedQuery(this.client, query, this.logger, this.cache, queryMetadata, cacheConfig, fields, name, isResponseInArrayMode, customResultMapper, false);
	}
	prepareRelationalQuery(query, fields, name, customResultMapper) {
		return new EffectPgPreparedQuery(this.client, query, this.logger, this.cache, void 0, void 0, fields, name, false, customResultMapper, true);
	}
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), void 0, void 0, false).execute();
	}
	all(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), void 0, void 0, false).all();
	}
	transaction(transaction) {
		const { dialect, relations, schema } = this;
		return this.client.withTransaction(effect_Effect.gen(this, function* () {
			return yield* transaction(new EffectPgTransaction(dialect, this, relations, schema));
		}));
	}
};
var EffectPgTransaction = class extends __pg_core_effect_session_ts.PgEffectTransaction {
	static [__entity_ts.entityKind] = "EffectPgTransaction";
};

//#endregion
exports.EffectPgPreparedQuery = EffectPgPreparedQuery;
exports.EffectPgSession = EffectPgSession;
exports.EffectPgTransaction = EffectPgTransaction;
